Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/preprocessor/BasicPreprocessor.java

```diff
@@ -0,0 +1,701 @@
+     * Since the state of the statistic is now changed based on more recent values
+     *
+     * @param gap              estimated value
+     * @param relativePosition how far back in the shingle
+     * @return transform of the time value to original input space
+     */
+    public long inverseMapTime(double gap, int relativePosition) {
+        // note this ocrresponds to differencing being always on
+        checkArgument(shingleSize + relativePosition >= 0, " error");
+        if (normalizeTime) {
+            return (long) Math.floor(previousTimeStamps[shingleSize - 1 + relativePosition]
+                    + timeStampDeviation.getMean() + 2 * gap * timeStampDeviation.getDeviation());
+        } else {
+            return (long) Math
+                    .floor(gap + previousTimeStamps[shingleSize - 1 + relativePosition] + timeStampDeviation.getMean());
+        }
+    }
+
+    /**
+     * determines the next point to fed to the model for a missing value
+     *
+     * @param fillin            strategy of imputation
+     * @param baseDimension     number of values to imput
+     * @param lastShingledPoint last full entry seen by model
+     * @return an array of baseDimension values corresponding to plausible next
+     *         input
+     */
+
+    protected double[] impute(RandomCutForest forest, ImputationMethod fillin, int baseDimension,
+            double[] lastShingledPoint) {
+        double[] result = new double[baseDimension];
+        if (fillin == ImputationMethod.ZERO) {
+            return result;
+        }
+        if (fillin == FIXED_VALUES) {
+            System.arraycopy(defaultFill, 0, result, 0, baseDimension);
+            return result;
+        }
+        int dimension = forest.getDimensions();
+        if (fillin == PREVIOUS) {
+            System.arraycopy(lastShingledPoint, dimension - baseDimension, result, 0, baseDimension);
+            return result;
+        }
+        if (forest.getTotalUpdates() < MINIMUM_OBSERVATIONS_FOR_EXPECTED || dimension < 4 || baseDimension >= 3) {
+            System.arraycopy(lastShingledPoint, dimension - baseDimension, result, 0, baseDimension);
+            return result;
+        }
+        int[] positions = new int[baseDimension];
+        double[] temp = Arrays.copyOf(lastShingledPoint, lastShingledPoint.length);
+        shiftLeft(temp, baseDimension);
+        for (int y = 0; y < baseDimension; y++) {
+            positions[y] = dimension - baseDimension + y;
+        }
+        double[] newPoint = forest.imputeMissingValues(temp, baseDimension, positions);
+        System.arraycopy(newPoint, dimension - baseDimension, result, 0, baseDimension);
+        return result;
+    }
+
+    /**
+     * applies transformations if desired
+     *
+     * @param inputPoint input point
+     * @return a differenced version of the input
+     */
+    protected double[] transformValues(double[] inputPoint, double[] factors) {
```

Comment:
This method makes it look like we should have a  preprocessor base class or interface with 4 implementations. BasicPreprocessor does not seem very basic!

Reply:
Fair point :) "There is no telling where the road takes us to ...". Fixing names.